### PR TITLE
Make GHC recognise our ($) operator.

### LINKED ISSUE #4555: Make GHC recognize our $ operator as he built-in $ operator
GHC has some special handling of the `$` operator, particularly around applicative `do`-notation. Unfortunately, our `$` (living in `daml-stdlib:DA.Internal.Prelude`) is not recognized to receive this special treatment. Thus, the following code requires a monad constraint although it is applicative:
```haskell
do
  x <- a
  y <- b
  return $ f x y
```
If you change the last line to `return (f x y)`, it works as intended.

To make this work with the `$` in place, we need to pass the check in
https://gitlab.haskell.org/ghc/ghc/blob/20484ae6e4bae6d992f68bc9e957f0804e26d1e4/compiler/rename/RnExpr.hs#L1940

### PR DESCRIPTION
**Pull request by sofiafaro-da**
Fixes #4555 by moving ($) to GHC.Base and giving it a levity polymorphic
signature, as GHC expects. Added a regression test.

changelog_begin
changelog_end

### Pull Request Checklist

- [ ] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/master/CONTRIBUTING.md)
- [ ] Include appropriate tests
- [ ] Set a descriptive title and thorough description
- [ ] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [ ] Include changelog additions in one or more commit message bodies between the `CHANGELOG_BEGIN` and `CHANGELOG_END` tags
- [ ] Normal production system change, include purpose of change in description

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

**Review by cocreature — APPROVED** [review]
Awesome, thank you!

File: compiler/damlc/daml-prim-src/GHC/Base.daml [hunk 1]
```diff
@@ -87,3 +89,11 @@ class Functor f where
     (<$)        =  fmap . const
 
 infixl 4  <$
+
+infixr 0 $
+-- | Take a function from `a` to `b` and a value of type `a`, and apply the
+-- function to the value of type `a`, returning a value of type `b`.
+-- This function has a very low precedence, which is why you might want to use
+-- it instead of regular function application.
+($) : forall r a (b : TYPE r). (a -> b) -> a -> b
```
Comment (hurryabit) [inline compiler/damlc/daml-prim-src/GHC/Base.daml:98]: Let's hope nobody has used `$` with explicit type applications so far. I mean, why would you, but still. And there's a simple workaround: just don't use `$`. 💸

**Review by hurryabit — COMMENTED** [review]
Very nice. Thank you.

File: compiler/damlc/daml-prim-src/GHC/Base.daml (hunk 1 above)
Comment (sofiafaro-da) [reply compiler/damlc/daml-prim-src/GHC/Base.daml:98]: So, technically this is breaking the existing standard library interface. But it's also a bug fix in the sense that it makes `($)` work with ApplicativeDo. Do you want to weigh in on this, @bame-da?

Comment (bame-da) [reply compiler/damlc/daml-prim-src/GHC/Base.daml:98]: All good from my end.

**Pull request merged**